eth_utils/abi.py

```diff
@@ -19,15 +19,15 @@ def collapse_if_tuple(abi):
     ... )
     '(address,uint256,bytes)'
     """
-    if not abi["type"].startswith("tuple"):
-        return abi["type"]
+    typ = abi["type"]
 
-    component_types = [collapse_if_tuple(component) for component in abi["components"]]
+    if not typ.startswith("tuple"):
+        return typ
 
-    collapsed = "(" + ",".join(component_types) + ")"
-
-    if abi["type"].endswith("[]"):
-        collapsed += "[]"
+    collapsed = "({delimited}){array_dims}".format(
+        delimited=",".join(collapse_if_tuple(c) for c in abi["components"]),
+        array_dims=typ[5:],
```

Comment:
I'm probably missing some context here, but why 5?

Reply:
I should maybe add a comment.  The string `"tuple"` is 5 characters long.  By that point in the function, we know the string begins with `"tuple"`.  According to the spec, a tuple type in an interface definition will have the form `tuple`, `tuple[]`, or `tuple[k]` where `k` is some constant.  So anything that follow the string "tuple" in a tuple type defines the array dimensions.